firedrake/variational_solver.py

```diff
@@ -92,7 +92,11 @@ def __init__(self, F, u, bcs=None, J=None,
             self.u_restrict = Function(V_res).interpolate(u)
             v_res, u_res = TestFunction(V_res), TrialFunction(V_res)
             F_arg, = F.arguments()
-            self.F = replace(F, {F_arg: v_res, self.u: self.u_restrict})
+            replace_F = {F_arg: v_res, self.u: self.u_restrict}
+            for c in F.coefficients():
+                if c.function_space() == V.dual():
+                    replace_F[c] = Cofunction(V_res.dual()).interpolate(c)
```

Comment:
I understand why you do this, but I would like to think a bit more to see if there is any better solution.

Reply:
I hoped something like `self.F = interpolate(F, V_res.dual())` would work, but it looks not supported, so maybe this is fine.

Reply:
I found a way to make `self.F = action(replace(F, {u: u_res}), interpolate(v_res, V))` work. It requires fixing some FormSum bugs, and fixing the way `assemble` optimizes and applies BCs on BaseForms.

Reply:
Sounds good.